solutions/task_4.py

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python3
+"""
+TASK: Implement classes "Temperature", "Person" using Python descriptor protocols to execute unittests without errors.
+"""
+import re
+import unittest
+from datetime import datetime
+
+
+class Temperature(object):
+
+    def __init__(self, value=0):
+        self.fahrenheit = value
+
+    def get_celsius(self):
+        return (self.fahrenheit - 32)/1.8
+
+    def set_celsius(self, value):
+        self.fahrenheit = value*1.8 + 32
+
+    celsius = property(get_celsius, set_celsius)
+
+
+class Descriptor(object):
+
+    def __init__(self, attr_name, attr_type):
+        self.attr_name = attr_name
+        self.attr_type = attr_type
+
+    def __get__(self, instance, owner):
+        return instance.__dict__[self.attr_name]
+
+    def __set__(self, instance, attr_value):
+        if self.attr_type != type(attr_value):
+            raise TypeError
+        if self.attr_name == 'phone':
+            pattern = r"[0-9]{3} [0-9]{2} [0-9]{7}$"
```

Comment:
```
pattern = r"[0-9]{3} [0-9]{2} [0-9]{7}$"
if re.match(pattern, attr_value):
```
![](http://memesmix.net/media/created/rkh3nk.jpg)

Reply:
I suppose to add ```^``` into pattern

Reply:
Actually it doesn't influence the outcome of the code but I think in terms of world there are some examples where the code of a country in a phone number consists of 2 or even 1 digits (+7 in Russia for example).

Reply:
1) ```^``` is not about digits number!
2) You have not tests for codes which contain more or less digits. You have tests for exactly phone/code length.